app/controllers/partners/family_requests_controller.rb

```diff
@@ -55,21 +60,55 @@ def validate
     private
 
     def build_family_requests_attributes(params)
-      children_ids = []
+      child_ids = params.keys.grep(/^child-/).map { |key| key.split('-').last }
 
-      params.each do |key, _|
-        is_child, id = key.split('-')
-        if is_child == 'child'
-          children_ids << id
-        end
-      end
+      validate_visible_items!(child_ids)
 
-      children = current_partner.children.where(id: children_ids).joins(:requested_items).select('children.*', :item_id)
+      children_grouped_by_item_id = current_partner
+        .children
+        .where(id: child_ids)
+        .joins(:requested_items)
+        .select('children.*', 'items.id as item_id',
+          'items.name as item_name',
+          'items.visible_to_partners')
+        .group_by(&:item_id)
 
-      children_grouped_by_item_id = children.group_by(&:item_id)
       children_grouped_by_item_id.map do |item_id, item_requested_children|
-        { item_id: item_id, person_count: item_requested_children.size, children: item_requested_children }
+        {
+          item_id: item_id,
+          person_count: item_requested_children.size,
+          children: item_requested_children
+        }
+      end
+    end
+
+    def validate_visible_items!(child_ids)
+      invisible_item_requests = current_partner
```

Comment:
You can reuse most if not all of the query that was created earlier. You can pass the query object into this method and just add the changes you need onto it.

Reply:
Having said that, all of this is starting to seem very heavy for controller logic. I think we should move most of this to the service class.